File: sdk/python/alp_sdk/arch_decomposer.py
from datetime import datetime, timezone
from typing import List, Dict, Any, Optional

class MonolithAnalysis:
    def __init__(self, target_path: str, total_files: int, coupling_score: float, modules: Dict[str, List[str]]):
        self.target_path = target_path
        self.total_files = total_files
        self.coupling_score = coupling_score
        self.modules = modules

class MicroservicePlan:
    def __init__(
        self,
        plan_id: str,
        monolith_target: str,
        proposed_services: List[str],
        service_boundaries: Dict[str, List[str]],
        coupling_score: float,
        created_at: Optional[str] = None,
    ):
        self.id = plan_id
        self.monolith_target = monolith_target
        self.proposed_services = proposed_services
        self.service_boundaries = service_boundaries
        self.coupling_score = coupling_score
        self.created_at = created_at or datetime.now(timezone.utc).isoformat()

class ArchDecomposerEngine:
    def analyze_monolith(self, target_path: str, file_paths: List[str]) -> MonolithAnalysis:
        modules: Dict[str, List[str]] = {
            "auth": [],
            "billing": [],
            "notifications": [],
            "core": [],
        }

        for f in file_paths:
            if "auth" in f or "user" in f or "token" in f:
                modules["auth"].append(f)
            elif "pay" in f or "stripe" in f or "billing" in f:
                modules["billing"].append(f)
            elif "email" in f or "push" in f or "notify" in f:
                modules["notifications"].append(f)
            else:
                modules["core"].append(f)

        non_core = len(modules["auth"]) + len(modules["billing"]) + len(modules["notifications"])
        score = max(0.1, min(0.95, round(1.0 - (non_core / max(1, len(file_paths))), 2)))

        return MonolithAnalysis(
            target_path=target_path,
            total_files=len(file_paths),
            coupling_score=score,
            modules=modules,
        )

    def decompose(self, analysis: MonolithAnalysis) -> MicroservicePlan:
        proposed: List[str] = []
        boundaries: Dict[str, List[str]] = {}

        for mod_name, files in analysis.modules.items():
            if len(files) > 0:
                svc_name = f"service-{mod_name}"
                proposed.append(svc_name)
                boundaries[svc_name] = files

        sanitized_id = "".join(c if c.isalnum() else "-" for c in analysis.target_path)

        return MicroservicePlan(
            plan_id=f"refactor-{sanitized_id}",
            monolith_target=analysis.target_path,
            proposed_services=proposed,
            service_boundaries=boundaries,
            coupling_score=analysis.coupling_score,
        )

File: parser/src/arch-decomposer.ts
export interface MonolithAnalysis {
  targetPath: string;
  totalFiles: number;
  couplingScore: number; // 0.0 (decoupled) to 1.0 (tightly coupled)
  modules: Record<string, string[]>;
}

export interface MicroservicePlan {
  id: string;
  monolithTarget: string;
  proposedServices: string[];
  serviceBoundaries: Record<string, string[]>;
  couplingScore: number;
  createdAt: string;
}

export class ArchDecomposerEngine {
  /**
   * Analyze monolith directory structure and dependency files into domain modules.
   */
  public analyzeMonolith(targetPath: string, filePaths: string[]): MonolithAnalysis {
    const modules: Record<string, string[]> = {
      auth: [],
      billing: [],
      notifications: [],
      core: [],
    };

    filePaths.forEach((file) => {
      if (file.includes('auth') || file.includes('user') || file.includes('token')) {
        modules.auth.push(file);
      } else if (file.includes('pay') || file.includes('stripe') || file.includes('billing')) {
        modules.billing.push(file);
      } else if (file.includes('email') || file.includes('push') || file.includes('notify')) {
        modules.notifications.push(file);
      } else {
        modules.core.push(file);
      }
    });

    const nonCoreCount = modules.auth.length + modules.billing.length + modules.notifications.length;
    const couplingScore = Math.max(0.1, Math.min(0.95, Number((1 - nonCoreCount / Math.max(1, filePaths.length)).toFixed(2))));

    return {
      targetPath,
      totalFiles: filePaths.length,
      couplingScore,
      modules,
    };
  }

  /**
   * Decompose MonolithAnalysis into discrete ALP `@package` microservices.
   */
  public decompose(analysis: MonolithAnalysis): MicroservicePlan {
    const proposedServices: string[] = [];
    const serviceBoundaries: Record<string, string[]> = {};

    Object.entries(analysis.modules).forEach(([modName, files]) => {
      if (files.length > 0) {
        const serviceName = `service-${modName}`;
        proposedServices.push(serviceName);
        serviceBoundaries[serviceName] = files;
      }
    });

    return {
      id: `refactor-${analysis.targetPath.replace(/[^a-zA-Z0-9]/g, '-')}`,
      monolithTarget: analysis.targetPath,
      proposedServices,
      serviceBoundaries,
      couplingScore: analysis.couplingScore,
      createdAt: new Date().toISOString(),
    };
  }
}
